### ISSUE TITLE
Get one of the previous layer by layer name (string)

### ISSUE THREAD

**Issue by zsdonghao** [tensorlayer/tensorlayer#512]
Hi all, I think we should support users to get one of the previous layer by using layer name, as follow:

```python
x = tf.placeholder(tf.float32, [None, 300])
net = InputLayer(x)
net = DenseLayer(net, name='dense1')
net = DenseLayer(net, name='dense2')

net['dense1'].outputs  <==== this one
```

### Reference

- [tl.layers.get_layers_with_name](http://tensorlayer.readthedocs.io/en/latest/modules/layers.html#get-layers-with-name)

- pycaffe

```python
net.tops['conv1'] = L.Convolution(bottom="image", kernel_size=3, stride=1, num_output=32, pad=1, weight_filler=dict(type='xavier'))
net.tops['relu1'] = L.ReLU(net.tops['conv1'], in_place=True)
net.tops['conv2'] = L.Convolution(net.tops['relu1'], kernel_size=3, stride=1, num_output=64, pad=1, weight_filler=dict(type='xavier'))
...

net.tops['conv1'] <==
```

**Comment by DEKHTIARJonathan** [comment]
@zsdonghao which API would you prefer using ?

```python
import tensorflow as tf
import tensorlayer as tl
x = tf.placeholder(tf.float32, [None, 300])
net = InputLayer(x)
net = DenseLayer(net, name='dense1')
net = DenseLayer(net, name='dense2')

## API 1:
layer = net['dense1']

## API 2:
layer = net.tops['dense1']
```

**Comment by zsdonghao** [comment]
@DEKHTIARJonathan I prefer to use API 1. Thanks, I added `help wanted` label ~

**Comment by DEKHTIARJonathan** [comment]
Here is some code to help on the implementation, the "Layer" class needs to be made **subscriptable**.

Here is a simple example on how to do this:



class SuperHero(object):
     def __init__(self):
         self.heros_strength= {
               "SuperMan": 666, 
               "Batman": 650, 
               "Hulk": 6e10, 
               "Prof. Xavier": 10
         }

     def __getitem__(self, item):
         return self.heros_strength[item]


myHero = SuperHero()

**Comment by BaptisteAmato** [comment]
I'll check this issue. I'm totally new to this repo so it'll help me dive into the code.

**Comment by BaptisteAmato** [comment]
It seems that in the reference [tl.layers.get_layers_with_name](http://tensorlayer.readthedocs.io/en/latest/modules/layers.html#get-layers-with-name), a Tensor is returned and not a Layer. Should "net['dense1']" return a Tensor? In that case, there could not be an output attribute "net['dense1'].outputs".

**Comment by DEKHTIARJonathan** [comment]
@zsdonghao I let you answer where you would like this contribution to be driven.

**Comment by zsdonghao** [comment]
@BaptisteAmato @DEKHTIARJonathan hi, for building a network, I think return a network could be better, otherwise, users have to feed the tensor into an `InputLayer` manually.

**Comment by DEKHTIARJonathan** [comment]
How would you implement this?
Before returning the Tensor you feed it into an InputLayer?
Because I'm not sure if there is anyway to find the previously created Layer object

**Comment by BaptisteAmato** [comment]
@zsdonghao I totally agree. However current code seems kind of weird to me: when using integer indexes, a network is returned like this:
```
net_new = Layer(prev_layer=None, name=self.name)
net_new.inputs = self.inputs
net_new.outputs = self.outputs[key]
net_new.all_layers = list(self.all_layers[:-1])
net_new.all_layers.append(net_new.outputs)
net_new.all_params = list(self.all_params)
net_new.all_drop = dict(self.all_drop)
return net_new
```
It seems that it returns the right outputs but the other parameters are from the entire network...

**Comment by zsdonghao** [comment]
@BaptisteAmato yes.. that is a problem .. 

What if we store all previous layer object here? https://github.com/tensorlayer/tensorlayer/blob/master/tensorlayer/layers/core.py#L408

**Comment by BaptisteAmato** [comment]
Yeah that's what I would do. @DEKHTIARJonathan  what do you think?

**Comment by DEKHTIARJonathan** [comment]
Could easily be done within the base class Layer.

**Comment by DEKHTIARJonathan** [comment]
You have two ways.

First, you create a list of layers and you loop through it.

However, names in TensorFlow shall be unique, so I would say that dictionary is a much more efficient approach.

One down side doing this, @zsdonghao we deprecated the function that managed tl layers name. Shall we bring it back to life?

If I have the time, I will have a look to Pytorch and look how they implemented this

**Comment by DEKHTIARJonathan** [comment]
I think we can used a linked list in Layer Base class. Each layer has a direct connection to the previous one (Layer) or ones (list of Layers: `ConcatLayer` & `MergeLayer`).
Usually a network does not have thousands of thousands of Layers. We should be fine about the time & space complexity.

**Comment by lgarithm** [comment]
I was going to add the **linked list** https://github.com/tensorlayer/tensorlayer/pull/469/files#diff-1a0995d02752855445936bb440663aceR401

**Comment by DEKHTIARJonathan** [comment]
What would you need to release to this feature @lgarithm?

**Comment by zsdonghao** [comment]
implementing here https://github.com/tensorlayer/tensorlayer/pull/755
